src/plugins/memdump/userhook.cpp

```diff
@@ -0,0 +1,671 @@
+        {
+            dll_meta->targets.emplace_back("CryptAcquireContextA", usermode_hook_cb, plugin);
+            dll_meta->targets.emplace_back("CryptAcquireContextW", usermode_hook_cb, plugin);
+        }
+        else if (strstr((const char*) dll_name->contents, "ws2_32.dll") != 0)
+        {
+            dll_meta->targets.emplace_back("WSAStartup", usermode_hook_cb, plugin);
+        }
+    }
+
+    drakvuf_release_vmi(drakvuf);
+    return !dll_meta->targets.empty();
+}
+
+/**
+ * Check if DLL is interesting, if so, build a "hooking context" of a DLL. Such context is needed,
+ * because user mode hooking is a stateful operation which requires a VM to be un-paused many times.
+ */
+static user_dll_t* create_dll_meta(drakvuf_t drakvuf, drakvuf_trap_info* info, memdump* plugin, addr_t dll_base)
+{
+    mmvad_info_t mmvad;
+    if (!drakvuf_find_mmvad(drakvuf, info->proc_data.base_addr, dll_base, &mmvad))
+        return nullptr;
+
+    if (mmvad.file_name_ptr == 0)
+        return nullptr;
+
+    auto vec_it = plugin->loaded_dlls.find(info->regs->cr3);
+
+    if (vec_it != plugin->loaded_dlls.end())
+    {
+        for (auto const& dll_meta : vec_it->second)
+        {
+            if (dll_meta.real_dll_base == mmvad.starting_vpn << 12)
+            {
+                PRINT_DEBUG("[MEMDUMP-USER] DLL %d!%llx is already hooked\n", info->proc_data.pid, (unsigned long long)mmvad.starting_vpn << 12);
+                return nullptr;
+            }
+        }
+    }
+
+    uint32_t thread_id;
+    if (!drakvuf_get_current_thread_id(drakvuf, info, &thread_id))
+        return nullptr;
+
+    user_dll_t dll_meta =
+    {
+        .dtb = info->regs->cr3,
+        .thread_id = thread_id,
+        .real_dll_base = (mmvad.starting_vpn << 12),
+        .is_hooked = false
+    };
+
+    if (!populate_hook_targets(drakvuf, plugin, mmvad, &dll_meta))
+        return nullptr;
+
+    PRINT_DEBUG("[MEMDUMP-USER] Found DLL which is worth processing %llx\n", (unsigned long long)mmvad.starting_vpn << 12);
+
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = mmvad.starting_vpn << 12
+    };
+
+    addr_t export_header_rva = 0;
+    size_t export_header_size = 0;
+
+    constexpr int MAX_HEADER_BYTES = 1024;   // keep under 1 page
+    uint8_t image[MAX_HEADER_BYTES];
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    bool status = (VMI_SUCCESS == peparse_get_image(vmi, &ctx, MAX_HEADER_BYTES, image));
+    drakvuf_release_vmi(drakvuf);
+
+    if (!status)
+        return nullptr;
+
+    void* optional_header = NULL;
+    uint16_t magic = 0;
+
+    peparse_assign_headers(image, NULL, NULL, &magic, &optional_header, NULL, NULL);
+    export_header_rva = peparse_get_idd_rva(IMAGE_DIRECTORY_ENTRY_EXPORT, &magic, optional_header, NULL, NULL);
+    export_header_size = peparse_get_idd_size(IMAGE_DIRECTORY_ENTRY_EXPORT, &magic, optional_header, NULL, NULL);
```

Comment:
Sanity check `rva` and `size`

Reply:
I've added two checks:

```
export_header_rva >= vad_length
export_header_size >= vad_length - export_header_rva
```

In normal situation a whole DLL (I belive) should belong to a single VAD node, so if the boundaries are exceeded then there is something abnormal. Hope this is correct.